rust/gitxetcore/src/stream/stdout_process_stream.rs

```diff
@@ -0,0 +1,403 @@
+        let Some(mut stdin) = process.stdin.take() else {
+            Err(anyhow!(
+                "Unable to connect stdin in child process {command:?}.",
+            ))?;
+            unreachable!();
+        };
+
+        stdin.write_all(stdin_data).await?; // Send all data
+        stdin.flush().await?; // Ensure all data is sent
+
+        Self::from_process(process, stdout_bufsize)
+    }
+
+    /// Constructs an Async Data Iterator reader that pipes the output from the stdout
+    /// of a process into a data iterator.
+    /// Reads things into an internal
+    /// buffer of a given size.
+    /// It is not guaranteed that every read will
+    /// fill the complete buffer.
+    pub fn from_process(mut child_process: tokio::process::Child, bufsize: usize) -> Result<Self> {
+        // Read the command's output from stdout in a streaming manner
+        let Some(stdout) = child_process.stdout.take() else {
+            Err(anyhow!("Unable to connect stdout in child process."))?;
+            unreachable!();
+        };
+
+        Ok(Self {
+            child_process: Some(child_process),
+            stdout,
+            bufsize,
+        })
+    }
+}
+
+#[async_trait]
+impl AsyncIterator<GitXetRepoError> for AsyncStdoutDataIterator {
+    type Item = Vec<u8>;
+
+    async fn next(&mut self) -> Result<Option<Self::Item>> {
+        let mut buffer = vec![0u8; self.bufsize];
+        let mut io_error = None;
+
+        loop {
+            match self.stdout.read(&mut buffer).await {
+                Ok(readlen) => {
+                    // If the readlen is 0, it could mean that the child process has paused, in which case
+                    // we simply want to wait and try again.  In this situation, coninue on to poll the child
+                    // process.
+                    if readlen > 0 {
+                        buffer.resize(readlen, 0);
+                        return Ok(Some(buffer));
+                    } else {
+                        break;
+                    }
+                }
+                Err(e) => {
+                    match e.kind() {
+                        ErrorKind::Interrupted | ErrorKind::WouldBlock => {
+                            // Simple retry
+                            continue;
+                        }
+                        ErrorKind::BrokenPipe | ErrorKind::UnexpectedEof => {
+                            // Done; check the exit status.
+                            break;
+                        }
+                        _ => {
+                            // Propogate all other errors.
+                            io_error = Some(e);
+                            break;
+                        }
+                    }
+                }
+            }
+        }
+
+        // First wait for the child process
+        let Some(child) = self.child_process.take() else {
+            Err(anyhow!("Next called after child process completed."))?;
```

Comment:
Maybe return EOF error instead of anyhow because this is still legit termination, so callers can check for this error.

Reply:
In this case, it would be a programming error -- EOF should always turn into None in the iterator, so they shouldn't be checking for EOF for that case.   This one only happens when the iterator gets reused after it sends a termination.